### ISSUE TITLE
[Feature request] Add support for selu activation to calculate_gain function.

### ISSUE THREAD

**Issue by hadaev8** [pytorch/pytorch#24991]
## 🚀 Feature

Add support for selu activation to calculate_gain function.

https://pytorch.org/docs/stable/_modules/torch/nn/init.html#calculate_gain

cc @albanD @mruberry @SsnL

**Comment by andreaskoepf** [comment]
@hadaev8 Do you know what the correct value has to be?

@t-vi reported last year a gain-value of 0.75 for selu was stable for him, see forums: https://discuss.pytorch.org/t/calculate-gain-tanh/20854/7

If I just look empirically at the std-gain I get the following numbers:

```
identity  : 1.0
sigmoid   : 4.809119701385498
tanh      : 1.5975916385650635
relu      : 1.7068026065826416
leaky_relu: 1.5421059131622314
selu      : 1.0002975463867188
elu       : 1.2694571018218994
```

If somebody is familiar with the relevant theory it would be great if he or she gave a hint how the numbers in `calculate_gain` were calculated. For example why has `torch.sigmoid` (Fermi function) a gain of 1?

Code used to calculate the std-gain:
```
x = torch.randn(100000)
def test_gain(x, fn):
    return x.std() / fn(x).std()

fns = [('identity', lambda x: x),
  ('sigmoid', torch.sigmoid), 
  ('tanh', torch.tanh),
  ('relu', torch.nn.functional.relu),
  ('leaky_relu', lambda x: torch.nn.functional.leaky_relu(x, 0.2)),
  ('selu', torch.nn.functional.selu),
  ('elu', torch.nn.functional.elu)]
result = [(name, test_gain(x, fn)) for name, fn in fns]
for name, gain in result:
  print('{:<10}: {}'.format(name, gain))
```

**Comment by hadaev8** [comment]
@andreaskoepf sorry, have no idea how this values work.

**Comment by andreaskoepf** [comment]
I can confirm the results reported by user jpeg729 and Thomas in the forums: https://discuss.pytorch.org/t/calculate-gain-tanh/20854/4 Their gain test-script is really nice and shows that the gradient can behave very differently from the layer activations .

Output of the test-script (100 layer deep net):

```
$ python init_gain_test.py selu 1
in: 1.0005
out: 1.0002 grad: 0.8287
out: 0.9998 grad: 1.1398
out: 0.9982 grad: 1.7574
out: 1.0030 grad: 2.3416
out: 1.0006 grad: 3.4042
out: 1.0037 grad: 4.5539
out: 0.9981 grad: 6.7281
out: 1.0047 grad: 9.2336
out: 1.0000 grad: 12.8418
out: 0.9986 grad: 18.7562

$ python init_gain_test.py selu 0.75
in: 1.0000
out: 0.7959 grad: 0.6368
out: 0.3066 grad: 0.2759
out: 0.2345 grad: 0.1911
out: 0.2074 grad: 0.2228
out: 0.2049 grad: 0.1948
out: 0.1884 grad: 0.2140
out: 0.1953 grad: 0.2590
out: 0.1974 grad: 0.2521
out: 0.2063 grad: 0.2637
out: 0.2141 grad: 0.2270
```

Regarding my code: 0 should be taken as the reference of the variance calculation since the weights of the following layer will project initially with a random sign (the sign of the activation function output does not matter, e.g. if it is biased).

Then the values are:
```
identity  : 1.0
sigmoid   : 1.8471702337265015
tanh      : 1.5913989543914795
relu      : 1.4173601865768433
leaky_relu: 1.3895978927612305
selu      : 1.000707983970642
elu       : 1.2470202445983887
```
The relu result then gets close to `math.sqrt(2)` (1.41421...) as expected. selu/elu values are not really affected that much.

Corrected test_gain() function:
```
def test_gain(x, fn):
    return x.pow(2).mean().sqrt() / fn(x).pow(2).mean().sqrt()
```

**Comment by hadaev8** [comment]
@andreaskoepf 
Are you gonna PR it?

**Comment by gchanan** [comment]
sounds good.

**Comment by hadaev8** [comment]
@gchanan 
Never contributed to such big repo, do I need anything besides adding activations and values to calculate_gain function?

**Comment by gchanan** [comment]
Hi @hadaev8: that plus add a test in test/test_nn.py (you can see some other calculate_gain examples in there.).

**Comment by mruberry** [comment]
Reserving for bootcamp.

**Comment by mruberry** [comment]
Releasing from bootcamp reservation.

**Comment by ajsanjoaquin** [comment]
This thread seems to have become stale. @andreaskoepf do you mind if I submit a PR for this? I was able to reproduce your gradient results with the selu function and the value 0.75.
```
in: 1.0008
out: 0.7968 grad: 0.6509
out: 0.3127 grad: 0.2760
out: 0.2404 grad: 0.2337
out: 0.2062 grad: 0.2039
out: 0.2056 grad: 0.1795
out: 0.2044 grad: 0.1977
out: 0.2005 grad: 0.2045
out: 0.2042 grad: 0.2273
out: 0.1944 grad: 0.2034
out: 0.2085 grad: 0.2464
```
But I'm not sure what do you use the std-gain values for. The gradients explode when putting in your std-gain value.
```
Using selu  with gain_value =1.000707983970642

in: 1.0000
out: 1.0011 grad: 0.8655
out: 1.0002 grad: 1.1315
out: 0.9985 grad: 1.6839
out: 1.0008 grad: 2.4107
out: 0.9999 grad: 2.9653
out: 1.0029 grad: 4.5212
out: 1.0068 grad: 6.8250
out: 1.0017 grad: 10.2219
out: 1.0026 grad: 13.3068
out: 1.0036 grad: 18.5051
```

I would like your clarification on this if possible. Otherwise, I'll set the selu gain_value to 0.75.

Also, I tried out the default gain value for tanh but the gradients explode (below). Can I change the tanh gain_value? Using a gain of 1 seems to maintain the gradients stable even at 100 layers. I can explore the other gain values too and add the necessary changes within the PR.
```
Using tanh with gain_value = 5/3
in: 1.0002
out: 0.7589 grad: 0.7070
out: 0.6514 grad: 1.7144
out: 0.6520 grad: 4.3221
out: 0.6501 grad: 11.4622
out: 0.6508 grad: 29.8123
out: 0.6517 grad: 76.7714
out: 0.6506 grad: 187.0547
out: 0.6508 grad: 498.8236
out: 0.6518 grad: 1332.0450
out: 0.6509 grad: 3348.0054
```
```
Using tanh with gain_value = 1
in: 1.0005
out: 0.6285 grad: 0.5451
out: 0.2188 grad: 0.1876
out: 0.1564 grad: 0.1455
out: 0.1282 grad: 0.1281
out: 0.1105 grad: 0.0968
out: 0.0976 grad: 0.1053
out: 0.0900 grad: 0.1133
out: 0.0831 grad: 0.0657
out: 0.0774 grad: 0.0602
out: 0.0722 grad: 0.0840
```

**Comment by hadaev8** [comment]
@andreaskoepf 
I'm using mish activation function (this why I'm ghosted).
Is it ok to add custom gain argument in pr?

**Comment by andreaskoepf** [comment]
@ajsanjoaquin @hadaev8 I never looked closer than what I originally  posted. I am not working on a PR. My test values only considered the forward gain (with 'artificial' 0 mean), not the gradient gain (which would be more interesting). A value of 0.75 for selu (as @t-vi found out) is better and should be used. I am also not aware what the other frameworks do and I would suggest to check this first.

**Comment by ajsanjoaquin** [comment]
@hadaev8 feel free to let me know if the PR needs editing. @gchanan may I request that you review my PR? And am I allowed to check & edit the other activations in custom_gain if they have useful gain values or do I need to open a new Issue for that? Thanks.

**Comment by hadaev8** [comment]
@ajsanjoaquin 
My only suggestion is custom gain value param in addition to the activation function name, so I can relay on native PyTorch with custom activation functions and reduce my codebase for a bit.

**Comment by ajsanjoaquin** [comment]
@hadaev8 Wouldn't it be easier to just explicitly declare a custom gain value for custom activation functions because you would write less? 

```python
import torch.nn.functional as F
a = torch.randn(1000,1000, requires_grad=True)
l = torch.nn.Linear(1000,1000, bias=False)
torch.nn.init.xavier_normal_(l.weight,10000)
F.custom_activation(torch.nn.Linear(1000,1000, bias=False)(a))
```
as opposed to
```python
a = torch.randn(1000,1000, requires_grad=True)
l = torch.nn.Linear(1000,1000, bias=False)
torch.nn.init.xavier_normal_(l.weight, torch.nn.init.calculate_gain("custom", 10000))
F.custom_activation(torch.nn.Linear(1000,1000, bias=False)(a))
```

Or am I misunderstanding your suggestion? Thanks.

**Comment by mrTsjolder** [comment]
Self-normalising networks are an alternative to normalisation layers. The goal is to induce a stable fixed point in the variance during the forward pass. To get this stable fixed point, the SELU activation function is used in combination with weights that have variance `1/fan_in`. It might be suboptimal for the backward pass, but the gain should definitely be 1 in order for the normalisation to work! The current implementation is wrong and makes it unnecessarily difficult to implement self-normalising networks.

**Comment by hadaev8** [comment]
@mrTsjolder What would you suggest for mish?
I seem to to have self-normalizing.
Same as selu linear gain?

**Comment by mrTsjolder** [comment]
> @mrTsjolder What would you suggest for mish?
> I seem to to have self-normalizing.
> Same as selu linear gain?

The moment propagation in swish-like functions is generally convex, which means that these induce unstable fixed points. The mish has this same property and can therefore not induce self-normalisation.
